### PR TITLE
Fix task handling bugs, so peers are more likely to be available

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Some of our crawler and connection rates can lead to weird interactions between crawlers, such as flapping or task synchronisation.

This is unexpected work in sprint 24, but it's related to the upcoming fanout and `notfound` work.

## Solution

Timer changes:
- Use prime numbers for crawler and heartbeat intervals, so they are almost always happening at different times
- Make sure each peer gets at least one connection attempt in each reconnection interval, even if the address book is full
- Add tests for relative timings
- Move some tests into their own modules

Task order changes:
- Let other tasks run between fanouts and retries, so we're more likely to choose different peers
- Let other tasks run after peer crawler DemandDrop, so we're more likely to get ready peers
- Yields are really difficult to test for, we'll replace this code soon in #2214 and #3235

## Review

@jvff or @dconnolly can review this PR.

This is urgent because syncs are very slow, and #3191 or #3244 should help fix that.

This PR is based on PR #3167, because they change the same import lines.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

- #3136
- #2214 
- #2726 
- #2156
- maybe #2153

### CODE REVIEW

**Comment by teor2345** [comment]
I've marked this as draft until PR #3167 merges, because they change the same import lines.

File: zebrad/src/components/mempool/crawler/tests/timing.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+use std::convert::TryInto;
+
+use zebra_chain::parameters::POST_BLOSSOM_POW_TARGET_SPACING;
+use zebra_network::constants::{DEFAULT_CRAWL_NEW_PEER_INTERVAL, HANDSHAKE_TIMEOUT};
+
+use crate::components::mempool::crawler::RATE_LIMIT_DELAY;
+
+#[test]
+fn ensure_timing_consistent() {
```
Comment (jvff) [inline zebrad/src/components/mempool/crawler/tests/timing.rs:11]: Nit: adding a verb to the name for extra clarity:

```suggestion
fn ensure_timings_are_consistent() {
```

**Review by jvff — APPROVED** [review]

**Comment by teor2345** [comment]
The Google Cloud build failed, I opened:
- #3239

**Comment by teor2345** [comment]
I've marked this PR and all its dependencies as a high priority, because this is a known bug that has actually happened before, and seems to still be happening.

I'm not sure if this PR fixes the bug, but it's an easy first step.

**Comment by mpguerra** [comment]
> I've marked this PR and all its dependencies as a high priority, because this is a known bug that has actually happened before, and seems to still be happening.
> 
> I'm not sure if this PR fixes the bug, but it's an easy first step.

@teor2345 should this still be in draft and are you looking for an additional review at this time? I see @jvff already approved a previous version of this PR a while ago.

**Review by conradoplg — DISMISSED** [review]
I believe this is just waiting for https://github.com/ZcashFoundation/zebra/pull/3167 to be merged and is a draft just to prevent it from merging before that one.

It looks good - I don't have much experience with this part of the code but was approved before, so I'm approving it now and leave it unmerged for a bit if anyone else wants to take a look.

**Review by jvff — DISMISSED** [review]

**Comment by conradoplg** [comment]
I rebased after merging https://github.com/ZcashFoundation/zebra/pull/3167, @jvff could you please check and reapprove?

**Comment by teor2345** [comment]
> I believe this is just waiting for #3167 to be merged and is a draft just to prevent it from merging before that one.

Yep, that's the only way to stop a GitHub PR that is based on another PR from merging into the wrong branch.

@mpguerra here's the process we agreed on a few months ago:
- if you finish a task or you're looking for work, check for PRs that need reviews (but you can skip draft PRs)
- if you want a review on a draft PR, ask for one, which I did: 
    - https://discord.com/channels/676527656170160146/716086297210650634/920797363885441054
    - https://discord.com/channels/676527656170160146/716086297210650634/921193869570367548
    - and by tagging people in the PR description

If there's something else I need to do, let's talk about it at the Engineering sync. And then let's document it in the onboarding doc.

**Comment by teor2345** [comment]
This got approved then rebased, and it's urgent, so I just admin-merged it.

**Pull request merged**